mono/eglib/gfile-unix.c

```diff
@@ -55,8 +55,26 @@ g_file_test (const gchar *filename, GFileTest test)
 	}
 
 	if ((test & G_FILE_TEST_IS_EXECUTABLE) != 0) {
+#if !defined(__PASE__)
 		if (access (filename, X_OK) == 0)
 			return TRUE;
+#else
+		/*
+		 * PASE always returns true for X_OK; contrary to how AIX
+		 * behaves (but *does* correspond to how it's documented!).
+		 * This behaviour is also consistent with the ILE, so it's
+		 * probably just an upcall returning the same results. As
+		 * such, workaround it.
+		 */
+		if (!have_stat)
+			have_stat = (stat (filename, &st) == 0);
+		if (have_stat && (st.st_mode & S_IXUSR) && (st.st_uid == getuid()))
+			return TRUE;
+		if (have_stat && (st.st_mode & S_IXGRP) && (st.st_gid == getgid()))
+			return TRUE;
+		if (have_stat && (st.st_mode & S_IXOTH))
+			return TRUE;
```

Comment:
This check is cheaper than the previous two. Do it before them?

Reply:
For that matter, nest everything inside `if (have_stat)`. I realize it perfs the same, but would be cleaner.

Reply:
```
	have_stat = have_stat || (stat (filename, &st) == 0); 

	if (have_stat) {
		if ((st.st_mode & S_IXOTH)
				|| ((st.st_mode & S_IXUSR) && (st.st_uid == getuid()))
				|| ((st.st_mode & S_IXGRP) && (st.st_gid == getgid()))
			return TRUE;
```